desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,1133 @@
+    intersect = intersect.reshape(P, S, -1)
+
+    # Only consider intersect if it is within knots that bound that polynomial.
+    is_intersect = ~jnp.isnan(intersect)
+    # Reorder so that all intersects along a field line are contiguous.
+    intersect = take_mask(intersect, is_intersect)
+    B_z_ra = take_mask(B_z_ra, is_intersect)
+    assert intersect.shape == B_z_ra.shape == (P, S, N * degree)
+    # Sign of derivative determines whether an intersect is a valid bounce point.
+    # Need to include zero derivative intersects to compute the WFB
+    # (world's fattest banana) orbit bounce integrals.
+    is_bp1 = B_z_ra <= 0
+    is_bp2 = B_z_ra >= 0
+    # The pairs bp1[i, j, k] and bp2[i, j, k] are boundaries of an integral only
+    # if bp1[i, j, k] <= bp2[i, j, k]. For correctness of the algorithm, it is
+    # required that the first intersect satisfies non-positive derivative. Now,
+    # because B_z_ra[i, j, k] <= 0 implies B_z_ra[i, j, k + 1] >= 0 by continuity,
+    # there can be at most one inversion, and if it exists, the inversion must be
+    # at the first pair. To correct the inversion, it suffices to disqualify the
+    # first intersect as a right boundary, except under the following edge case.
+    edge_case = (B_z_ra[..., 0] == 0) & (B_z_ra[..., 1] < 0)
+    is_bp2 = put_along_axis(is_bp2, jnp.array(0), edge_case, axis=-1)
+    # Get ζ values of bounce points from the masks.
+    bp1 = take_mask(intersect, is_bp1)
+    bp2 = take_mask(intersect, is_bp2)
+    # Consistent with (in particular the discussion on page 3 and 5 of)
+    # V. V. Nemov, S. V. Kasilov, W. Kernbichler, M. F. Heyn.
+    # Evaluation of 1/ν neoclassical transport in stellarators.
+    # Phys. Plasmas 1 December 1999; 6 (12): 4622–4632.
+    # https://doi.org/10.1063/1.873749.
+    # we ignore the bounce points of particles assigned to a class that
+    # are trapped outside this snapshot of the field line. The caveat
+    # is that the field line discussed in the paper above specifies the
+    # flux surface completely as its length tends to infinity, whereas
+    # the field line snapshot here is for a particular alpha coordinate.
+    # Don't think it's necessary to stitch together the field lines using
+    # rotational transform to potentially capture the bounce point outside
+    # this snapshot of the field line.
+
+    if check:
+        msg = "Bounce points have an inversion. Maybe create an issue on GitHub."
+        assert not jnp.any(bp1 > bp2), msg
+        msg = "Discontinuity detected. Is B_z_ra the derivative of the spline of B?"
+        assert not jnp.any(bp1[..., 1:] < bp2[..., :-1]), msg
+
+    return bp1, bp2
+
+
+def _affine_bijection_forward(x, a, b):
+    """[a, b] ∋ x ↦ y ∈ [−1, 1]."""
+    y = 2 * (x - a) / (b - a) - 1
+    return y
+
+
+def affine_bijection_reverse(x, a, b):
+    """[−1, 1] ∋ x ↦ y ∈ [a, b]."""
+    y = (x + 1) / 2 * (b - a) + a
+    return y
+
+
+def grad_affine_bijection_reverse(a, b):
```

Comment:
I think you call _grad_affine_bijection in the tests/test_bounce_integral which is going to throw an error.